darshan-util/pydarshan/darshan/experimental/plots/data_access_by_filesystem.py

```diff
@@ -0,0 +1,632 @@
+                                                df_writes=df_writes)
+    # if either of the Series are effectively empty we want
+    # to produce a new Series with the filesystem_root indices
+    # and count values of 0 (for plotting purposes)
+    read_groups, write_groups = check_empty_series(read_groups=read_groups,
+                                                   write_groups=write_groups,
+                                                   filesystem_roots=filesystem_roots)
+    if verbose:
+        print("read_groups:\n", read_groups)
+        print("write_groups:\n", write_groups)
+    return (read_groups, write_groups)
+
+
+def plot_data(fig: Any,
+              file_rd_series,
+              file_wr_series,
+              bytes_rd_series,
+              bytes_wr_series,
+              filesystem_roots: Sequence[str],
+              num_cats: Optional[int] = None):
+    """
+    Produce the horizontal bar plots for the data
+    access by category analysis.
+
+    Parameters
+    ----------
+
+    fig: matplotlib ``figure`` object into which the
+    bar plots will be placed
+
+    file_rd_series: a ``pd.Series`` object with per-category
+    counts of unique files from which at least 1 byte has been
+    read
+
+    file_wr_series: a ``pd.Series`` object with per-category
+    counts of unique files to which at least 1 byte has been
+    written
+
+    bytes_rd_series: a ``pd.Series`` object with per-category
+    counts of bytes read
+
+    bytes_wr_series: a ``pd.Series`` object with per-category
+    counts of bytes written
+
+    filesystem_roots: a sequence of strings containing unique filesystem root paths
+
+    num_cats: an integer representing the number of categories
+    to plot; default ``None`` plots all categories
+    """
+    list_byte_axes: list = []
+    list_count_axes: list = []
+    # use log10 scale if range exceeds
+    # two orders of magnitude in a column
+    use_log = [False, False]
+    for idx, series_pair in enumerate([[bytes_rd_series, bytes_wr_series, 1048576],
+                                       [file_rd_series, file_wr_series, 1]]):
+        maxval = max(series_pair[0].max(), series_pair[1].max())
+        minval = max(min(series_pair[0].min(), series_pair[1].min()), 1)
+        # adjust ratio to MiB when needed
+        ratio = ((maxval / series_pair[2]) / (minval / series_pair[2]))
+        if ratio > 100:
+            use_log[idx] = True
+
+    if num_cats is None:
+        num_cats = len(filesystem_roots)
+
+    for row, filesystem in enumerate(filesystem_roots[:num_cats]):
+        ax_filesystem_bytes = fig.add_subplot(len(filesystem_roots),
+                                              2,
+                                              row * 2 + 1)
+        ax_filesystem_counts = fig.add_subplot(len(filesystem_roots),
+                                              2,
+                                              row * 2 + 2)
```

Comment:
Since you are taking the first `num_cats` filesystem roots, I think you have to adjust the subplots here to:

```python
        ax_filesystem_bytes = fig.add_subplot(len(filesystem_roots[:num_cats]),
                                              2,
                                              row * 2 + 1)
        ax_filesystem_counts = fig.add_subplot(len(filesystem_roots[:num_cats]),
                                              2,
                                              row * 2 + 2)
```

or the subplots are not going to scale correctly. For example, if you have 20 filesystems and you only want the top 5, it will create subfigures as if there are 20 and only fill 1/4 of the space. I observed this locally when setting `num_cats=8` for `nonmpi_partial_modules.darshan`.